solver/types.go

```diff
@@ -73,6 +73,7 @@ type CachedResult interface {
 }
 
 type ResultProxy interface {
+	Clone() ResultProxy
```

Comment:
I'm a bit bothered by the change of this interface.

Let's say there is a different implementation of `ResultProxy`. The `Clone()` is not really part of that implementation. It is just a reference counting on top of it that doesn't change if the actual implementation changes.

Instead of calling `.Clone()` I think the code should add a private wrapper around it that implements own `Release()` but leaves other methods as passthrough.

smth like

```
type ref struct {
	main      sharedResultProxy
}

type sharedResultProxy {
     solver.ResultProxy
     ... // ref count fields
}

func (srp *sharedResultProxy) clone()
```

Reply:
I think having a private `clone()` method means that we have to duplicate that logic in both the `gateway` and `forward` packages.

I agree though, I like the idea of isolating the ref counting. Instead of going for a `sharedResultProxy`, I like the idea of using a `resultProxySplit`, similar to how we have `splitResult` in https://github.com/moby/buildkit/blob/v0.10/solver/result.go#L43-L47, with a public helper function like `SplitResultProxy() (ResultProxy, ResultProxy)`. Details of GC are then entirely in the splitting, and we avoid needing to have duplication of `clone`.